### PR TITLE
EOS-8251: cas: move meta ctg unlock for non-meta ops

### PR DESCRIPTION
**Pull request by sergey-shilov**
Problem:
Current implementation of CAS service contains
a possible two-factor (CAS + BE) dead lock
occurred during simultaneous component catalogue
deletion and putting key/value into that catalogue.

Dead lock model is the following:

 * Fom X tries to put key/val into catalogue A_cat,
   it write-locked the lock A_lock
 * Fom Y tries to delete catalogue A_cat, it tries
   to write-lock A_lock
 * Fom X has a tx in state M0_BTS_GROUPING
 * Fom Y has a tx in state M0_BTS_ACTIVE
 * Grouping can not be finished while there is an
   active tx, that can not be closed due to waiting
   on A_lock

So CAS locks are held and BE engine is "frozen".

Solution:
Currently meta ctg is unlocked right after meta
lookup is done. We can hold this read-lock longer
and unlock when TXN becomes active so that the
openning of the TXN is protected by meta read-lock,
in this case meta operation will wait until TXN is
active for non-meta operation and BE engine and CAS
will not get stuck.

Warning:
Solution may affect the performance.

Signed-off-by: Sergey Shilov <sergey.shilov@seagate.com>

### CODE REVIEW

**Review by just-now — APPROVED** [review]
This patch potentially may effect performance and stability cases therefore needs to be tested thoroughly. Here we take meta lock to take another "lock" or resource (opened transaction).
Eventually, a proper solution has to be introduced.

**Comment by cortx-admin** [comment]
<h1>Jenkins CI Result : <a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2/>cortx-motr#2</a></h1><h2>Motr Test Summary</h2><table><tr><th>Test Result</th><th>Count</th><th>Info</tr><tr><td>:x:Failed</td><td>0</td><td><details><summary>:file_folder:</summary><p></p></details></td></tr><tr><td>:checkered_flag:Skipped</td><td>5</td><td><details><summary>:file_folder:</summary><p><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/31motr-sys-kvs-kernel-test target=_blank>31motr-sys-kvs-kernel-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/07multi-client-tests target=_blank>07multi-client-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/39mount-fail target=_blank>39mount-fail</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/12sns-cc-io-tests target=_blank>12sns-cc-io-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/37protocol target=_blank>37protocol</a></p></details></td></tr><tr><td>:heavy_check_mark:Passed</td><td>57</td><td><details><summary>:file_folder:</summary><p><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/40motr-dgmode target=_blank>40motr-dgmode</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/51motr-rmw target=_blank>51motr-rmw</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/50motr-rm-lock-cc-io target=_blank>50motr-rm-lock-cc-io</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/20pool-version-test target=_blank>20pool-version-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/46m0crate target=_blank>46m0crate</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/54sns-repair-motr-1f target=_blank>54sns-repair-motr-1f</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/25spiel-sns-repair-test target=_blank>25spiel-sns-repair-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/21spiel-test target=_blank>21spiel-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/59sns-repair-motr-abort target=_blank>59sns-repair-motr-abort</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/01kernel-tests target=_blank>01kernel-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/33motr-st target=_blank>33motr-st</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/42motr-utils target=_blank>42motr-utils</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/47motr-tests-user-kernel target=_blank>47motr-tests-user-kernel</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/19sns-abort-test target=_blank>19sns-abort-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/09poolmach-tests target=_blank>09poolmach-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/13sns-repair-quiesce target=_blank>13sns-repair-quiesce</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/41motr-conf-update target=_blank>41motr-conf-update</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/32m0t1fs-rconfc-fail-test target=_blank>32m0t1fs-rconfc-fail-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/17sss-st-test target=_blank>17sss-st-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/00userspace-tests target=_blank>00userspace-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/03initscript-tests target=_blank>03initscript-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/45motr-sns-repair-N-1 target=_blank>45motr-sns-repair-N-1</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/30hagen-test target=_blank>30hagen-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/11sns-multi-tests target=_blank>11sns-multi-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/58sns-repair-motr-ios-fail target=_blank>58sns-repair-motr-ios-fail</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/29confgen-test target=_blank>29confgen-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/34sns-repair-1n-1f target=_blank>34sns-repair-1n-1f</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/44motr-sns-repair target=_blank>44motr-sns-repair</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/08degraded-mode-tests target=_blank>08degraded-mode-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/26spiel-sns-repair-quiesce-test target=_blank>26spiel-sns-repair-quiesce-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/60sns-repair-motr-abort-quiesce target=_blank>60sns-repair-motr-abort-quiesce</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/35m0singlenode target=_blank>35m0singlenode</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/14m0t1fs-fsync-test target=_blank>14m0t1fs-fsync-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/43motr-sync-replication target=_blank>43motr-sync-replication</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/04net-tests target=_blank>04net-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/55sns-repair-motr-1n-1f target=_blank>55sns-repair-motr-1n-1f</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/24m0d-fsync-test target=_blank>24m0d-fsync-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/49motr-rpc-cancel target=_blank>49motr-rpc-cancel</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/56sns-repair-motr-mf target=_blank>56sns-repair-motr-mf</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/38sns-abort-quiesce-test target=_blank>38sns-abort-quiesce-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/15m0t1fs-fwait-test target=_blank>15m0t1fs-fwait-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/02system-tests target=_blank>02system-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/22rpc-cancel-test target=_blank>22rpc-cancel-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/53clusterusage-alert target=_blank>53clusterusage-alert</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/05rpcping-tests target=_blank>05rpcping-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/23m0d-signal-test target=_blank>23m0d-signal-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/28motr-sys-kvs-test target=_blank>28motr-sys-kvs-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/06console-tests target=_blank>06console-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/18m0mount-test target=_blank>18m0mount-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/27sns-repair-io-fail-test target=_blank>27sns-repair-io-fail-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/16conf-st-test target=_blank>16conf-st-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/52singlenode-sanity target=_blank>52singlenode-sanity</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/57sns-repair-motr-1k-1f target=_blank>57sns-repair-motr-1k-1f</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/48motr-raid0-io target=_blank>48motr-raid0-io</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/10sns-single-tests target=_blank>10sns-single-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/36spare-reservation target=_blank>36spare-reservation</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node/51kem target=_blank>51kem</a></p></details></td></tr><tr><td>Total</td><td>62</td><td><a href=http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/Cortx-PR-Test/job/cortx-motr/2//testReport/junit/motr-single-node>:link:</a></td></tr></table><h2> CppCheck Summary</h2><p> &emsp;&emsp;&emsp;Cppcheck: No new warnings found :thumbsup:</p>

File: cas/service.c [hunk 1]
```diff
@@ -2550,6 +2561,14 @@ static struct m0_sm_state_descr cas_fom_phases[] = {
 		.sd_name      = "prep",
 		.sd_allowed   = M0_BITS(M0_FOPH_TXN_OPEN, M0_FOPH_FAILURE)
 	},
+	[CAS_TXN_OPENED] = {
+		.sd_name      = "txn-opened",
+		.sd_allowed   = M0_BITS(CAS_META_UNLOCK, CAS_LOOP)
+	},
+	[CAS_META_UNLOCK] = {
+		.sd_name      = "meta_unlock",
```
Comment (imvenkip) [inline cas/service.c:2569]: s/meta_unlock/meta-unlock

File: cas/service.c [hunk 2]
```diff
@@ -1330,6 +1333,14 @@ static int cas_fom_tick(struct m0_fom *fom0)
 		 * fixed for the correct credit calculation.
 		 */
 		break;
+	case CAS_TXN_OPENED:
+		m0_fom_phase_set(fom0, is_meta ? CAS_LOOP : CAS_META_UNLOCK);
+		break;
+	case CAS_META_UNLOCK:
+		M0_ASSERT(!is_meta);
+		m0_long_read_unlock(m0_ctg_lock(meta), &fom->cf_meta);
```
Comment (huanghua78) [inline cas/service.c:1341]: For meta, when does unlock happen?

Comment (sergey-shilov) [reply cas/service.c:1341]: For meta operations _fom->cf_ctg_ points to meta catalogue, it's released when operation is done.

**Review by nikitadanilov — APPROVED** [review]

**Pull request merged**